### ISSUE TITLE
debian-10-minimal requires 'gcr' package for sys-firewall GUI

### ISSUE THREAD

**Issue by keypoc** [QubesOS/qubes-issues#6381]
**Qubes OS version**

4.0

**Affected component(s) or functionality**

Firewall GUI

**Brief summary**

I am getting an error on the debian-10-minimal based sys-firewall GUI.

**How Reproducible**

Instantly

**To Reproduce**

1) install 'qubes-core-agent-networking' in TemplateVM debian-10-minimal (clean install)
2) create sys-firewall based on this template
3) connect any AppVM to this firewall
4) run 'Qube Settings' to AppVM and go to the 'Firewall Rules' tab 

**Expected behavior**

All required functionality should be working by default when qubes-core-agent-networking is installed

**Actual behavior**

I get an error

```
The 'personal' qube is network connected to 'firewall', which does not support firewall!
You may edit the 'test' qube firewall rules, but these will not take any effect until you connect it to a working Firewall qube.
```

**Solutions you've tried**

Installing the 'gcr' package in TemlateVM, which sys-firewall is based on, resolves this error.

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

https://www.qubes-os.org/doc/templates/minimal/

**Comment by andrewdavidwong** [comment]
Please note the following with respect to minimal template expectations:

https://www.qubes-os.org/doc/templates/minimal/#important

**Comment by unman** [comment]
On Tue, Feb 02, 2021 at 01:49:51AM -0800, Andrew David Wong wrote:
> Please note the following with respect to minimal template expectations:
> 
> https://www.qubes-os.org/doc/templates/minimal/#important
> 

True indeed - but in this case, it's not unreasonable to expect it to
work out of the box - there's clearly something missing in the
dependencies of the networking package for Debian.
I'll look in to it.

**Comment by keypoc** [comment]
> On Tue, Feb 02, 2021 at 01:49:51AM -0800, Andrew David Wong wrote: Please note the following with respect to minimal template expectations: https://www.qubes-os.org/doc/templates/minimal/#important
> True indeed - but in this case, it's not unreasonable to expect it to work out of the box - there's clearly something missing in the dependencies of the networking package for Debian. I'll look in to it.

For example sys-net does not have this GUI error. In 'qubes-core-agent-network-manager' contains the 'gnome-keyring' package, which depends on the 'gcr library'.
The most interesting thing is that when you subsequently remove the 'gcr' package, the error is no longer reproduced in the gui sys-firewall. Magic.

**Comment by SvenSemmler** [comment]
> The most interesting thing is that when you subsequently remove the 'gcr' package, the error is no longer reproduced in the gui sys-firewall. Magic.

I am not sure all conclusions here are correct. All my qubes are based on debian-minimal and I am very familiar with the error message the reporter encountered. It was also brought up multiple times in qubes-users:

https://www.mail-archive.com/qubes-users@googlegroups.com/msg32374.html
https://qubes-os.discourse.group/t/debian-10-minimal-firewall-issues/146/8

I think I have found the root cause and documented it here: https://qubes-os.discourse.group/t/debian-10-minimal-firewall-issues/146/9?u=sven

Basically the *template* which sys-firewall is based upon requires qvm-feature qubes-firewall 1, because that's what the GUI checks for some reason. 

@keypoc  my firewall works, I don't see the error and 'gcr' is not installed. Could you please cross-check your findings against what I found?

**Comment by keypoc** [comment]
> > The most interesting thing is that when you subsequently remove the 'gcr' package, the error is no longer reproduced in the gui sys-firewall. Magic.
> 
> I am not sure all conclusions here are correct. All my qubes are based on debian-minimal and I am very familiar with the error message the reporter encountered. It was also brought up multiple times in qubes-users:
> 
> https://www.mail-archive.com/qubes-users@googlegroups.com/msg32374.html
> https://qubes-os.discourse.group/t/debian-10-minimal-firewall-issues/146/8
> 
> I think I have found the root cause and documented it here: https://qubes-os.discourse.group/t/debian-10-minimal-firewall-issues/146/9?u=sven
> 
> Basically the _template_ which sys-firewall is based upon requires qvm-feature qubes-firewall 1, because that's what the GUI checks for some reason.
> 
> @keypoc my firewall works, I don't see the error and 'gcr' is not installed. Could you please cross-check your findings against what I found?

I was able to reproduce your problem and solution.
minimal clean debian install
1) Installed 'qubes-core-agent-networking'
2) Created a sys-firewall based on this template
3) Connected app-vm to this firewall
4) installed 'qubes-firewall 1' on sys-firewall
`'qvm-features sys-firewall 'qubes-firewall 1'`
6) checked dom0 again and checked for error

In fact, the error is gone and 'qubes-firewall 1' shows up as enabled in 'qvm-features'.
I tested the standard system firewall with a clean install of qubes without 'qubes-firewall 1' enabled and did not notice that it was set to mode '1'. That is, if it does not appear when prompted for 'qvm-features', then it is working by default.
Based on the information provided on
https://dev.qubes-os.org/projects/core-admin-client/en/latest/manpages/qvm-features.html,
namely:
```
This command is used to manually manage domain functions. Functions are key-value pairs, with key and value being strings. They are used by extensions to store information about the domain and make policy decisions based on them. For example, they can indicate that a particular software package has been installed within the template and the domains based on it have certain capabilities.

Warning Functions are usually controlled by the extensions themselves and you should not modify them directly. Otherwise, strange things can happen.
Some extensions interpret values ​​as boolean. In this case, an empty string means False, and a non-empty string (usually "1") means True. No function means "default" which depends on the extension. In most cases, the default value for the function is derived from the qube template.
```

If I understand correctly, then TemplateVM is checked for the presence of the necessary packages for a specific service, and in fact everything should work by default without enabling 'qubes-firewall 1', if the required package is available.

I am guessing this is due to the lack of a required package in 'qubes-core-agent-networking'. For example, if you additionally install 'qubes-core-agent-network-manager' or simple 'gnome-keyring', then this error will disappear. Personally, I traced this back to the 'gcr' package by installing each package separately and committing the bug fix.
Try to reproduce my steps and you will see that the solution to this problem is to simply install the 'gcr' package without enabling 'qvm-features qubes-firewall 1'.

**Comment by SvenSemmler** [comment]
> If I understand correctly, then TemplateVM is checked for the
> presence of the necessary packages for a specific service, and in
> fact everything should work by default without enabling
> 'qubes-firewall 1', if the required package is available.

Which in fact it does... the firewall works when 
qubes-core-agent-networking is installed, only the Qube Manager reports 
the error based on the feature flag at the template.

> Try to reproduce my steps and you will see that the
> solution to this problem is to simply install the 'gcr' package
> without enabling 'qvm-features qubes-firewall 1'.

Did installing 'gcr' set the qubes-firewall feature?

The code in Qube Manager:

if netvm is not None and
not netvm.features.check_with_template(
‘qubes-firewall’,
False):

... which means that if there is a NetVM but the feature is not set 
you'll see the error message.

I will try an reproduce your findings later in the evening, but also 
hope that @unman will be able to shed some more light on this after 
looking into it.

Thank you @keypoc!

**Comment by keypoc** [comment]
> Did installing 'gcr' set the qubes-firewall feature?
Yes

TemplateVM with 'qubes-core-agent-networking' installed displays at the request 'qvm-features debian-10-minimal' displays: (**here qubes-firewall is clearly in 'false' mode - that is, disabled**)
```
check-updates                          
supported-service.crond                1
qrexec                                 1
supported-service.updates-proxy-setup  1
gui                                    1
supported-service.meminfo-writer       1
qubes-firewall                         
supported-service.clocksync            1
supported-service.qubes-update-check   1
updates-available                      
os                                     Linux
```

After installing 'gcr' in TemplateVM:
```
check-updates                          
supported-service.crond                1
qrexec                                 1
supported-service.updates-proxy-setup  1
gui                                    1
supported-service.meminfo-writer       1
qubes-firewall                         1
supported-service.clocksync            1
supported-service.qubes-update-check   1
updates-available                      
supported-service.qubes-firewall       1
supported-service.qubes-network        1
supported-service.qubes-updates-proxy  1
os                                     Linux
```
As a result, qubes-firewall has mode "1" - true
Also, two additional functions are now available:
```
supported-service.qubes-network        1
supported-service.qubes-updates-proxy  1
```
This suggests that the 'gcr' package activates the mode 'qubes-firewall 1' feature and two more. But this in no way explains the fact that 'qubes-firewall' itself works inside appvm if 'qubes-firewall' is in 'false' mode, when I request 'systemctl status qubes-firewall', I see no problems, and also when looking at iptables/nftables rules. Well, at least I think it works. Only this error is in the GUI.
The funny thing is that when you uninstall the 'gcr' package, the error no longer appears, as if this package is not needed =) The output of 'qvm-features' is the same as after installing 'gcr'.

**Comment by SvenSemmler** [comment]
> as if this package is not needed

That is my suspicion.

I think the flag should be set by qubes-core-agent-networking and not by 
gcr.

**Comment by unman** [comment]
On Tue, Feb 02, 2021 at 02:58:03PM -0800, Sven Semmler wrote:
> 
> > as if this package is not needed
> 
> That is my suspicion.
> 
> I think the flag should be set by qubes-core-agent-networking and not by 
> gcr.
> 

Undoubtedly true

**Comment by ideologysec** [comment]
does manually running `qvm-features <vm-name> qubes-firewall 1` in dom0 allow for the firewall to work without installing gcr? (similar bug affects mirage-firewall and needs to be manually set)
